### PR TITLE
Pass --warn-error to Odoc with (env ..)

### PR DESCRIPTION
**Pull request by Julow**
Alternative PR to https://github.com/ocaml/dune/pull/3018

This PR adds a way to pass `--warn-error` to Odoc by adding a field to the `env` stanza:

```
(env
 (<profile>
  (odoc
   (warnings fatal))))
```

Warnings can be set to `fatal` or `nonfatal`. They are `nonfatal` by default (for now).

Enabling this option and having an Odoc version that doesn't support it is an error. (no attempt is made to catch this)

### CODE REVIEW

**Comment by diml** [comment]
> Enabling this option and having an Odoc version that doesn't support it is an error. (no attempt is made to catch this)

Just update the odoc conflict in the `dune-project` file. No need to make life complicated :)

File: src/dune/dune_env.ml [hunk 1]
```diff
@@ -105,6 +121,10 @@ module Stanza = struct
           User_error.raise ~loc
             [ Pp.textf "Variable %s is specified several times" k ] )
 
+  let odoc_field =
+    field "odoc" ~default:Odoc.empty
+      (Dune_lang.Syntax.since Stanza.syntax (2, 0) >>> Odoc.decode)
```
Comment (diml) [inline src/dune/dune_env.ml]:
```suggestion
      (Dune_lang.Syntax.since Stanza.syntax (2, 2) >>> Odoc.decode)
```

Since this will be part of Dune 2.2. You also need to update the variable `latest_version` to `(2, 2)` in `stanza.ml`

File: src/dune/env_node.ml [hunk 2]
```diff
@@ -148,3 +154,24 @@ let rec foreign_flags t ~profile ~expander ~default_context_flags =
...
     t.foreign_flags <- Some flags;
     flags
+
+let rec odoc t ~profile =
+  match t.odoc with
+  | Some x -> x
+  | None ->
+    let open Odoc in
+    let default =
+      (* TODO: Enable for dev profile in the future *)
```
Comment (diml) [inline src/dune/env_node.ml]:
```suggestion
      (* DUNE3: Enable for dev profile in the future *)
```

This way we won't miss it when we switch to Dune 3.0.0.

File: test/blackbox-tests/test-cases/odoc/warn-error/run.t [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  (cd _build/default/_doc/_odoc/pkg/foo && %OPAM%/bin/odoc compile --warn-error --pkg foo -o page-index.odoc ../../../_mlds/foo/index.mld)
+  (cd _build/default/lib/.foo.objs/byte && %OPAM%/bin/odoc compile --warn-error -I . -I ../../../_doc/_odoc/pkg/foo --pkg foo -o foo.odoc foo.cmti)
+  (cd _build/default/_doc/_html && %OPAM%/bin/odoc html --warn-error -I ../_odoc/pkg/foo -I ../../lib/.foo.objs/byte -o . ../_odoc/pkg/foo/page-foo.odoc)
+  (cd _build/default/_doc/_html && %OPAM%/bin/odoc html --warn-error -I ../_odoc/pkg/bar -I ../../sub_env/lib/.bar.objs/byte -o . ../_odoc/pkg/bar/page-index.odoc)
+  (cd _build/default/_doc/_html && %OPAM%/bin/odoc html -I ../_odoc/pkg/bar -I ../../sub_env/lib/.bar.objs/byte -o . ../../sub_env/lib/.bar.objs/byte/bar.odoc)
+  (cd _build/default/_doc/_html && %OPAM%/bin/odoc html --warn-error -I ../_odoc/pkg/foo -I ../../lib/.foo.objs/byte -o . ../_odoc/pkg/foo/page-index.odoc)
+  (cd _build/default/_doc/_html && %OPAM%/bin/odoc html --warn-error -I ../_odoc/pkg/foo -I ../../lib/.foo.objs/byte -o . ../../lib/.foo.objs/byte/foo.odoc)
```
Comment (diml) [inline test/blackbox-tests/test-cases/odoc/warn-error/run.t]: This output feels a bit too verbose. In order to convince ourselves that the test is passing, we have to go through it to check that `--warn-error` is correctly passed where it should. What about actually leaving an odoc error and checking that it is reported as an error?

**Review by ghost — CHANGES_REQUESTED** [review]

File: test/blackbox-tests/test-cases/odoc/warn-error/run.t (hunk 3 above)
Comment (Julow) [reply test/blackbox-tests/test-cases/odoc/warn-error/run.t]: Generating a warning from the `odoc html` is not easy and Odoc is called from multiple independent paths in the source code and I wasn't convinced each paths had the correct environment.
Would it be better if I do both ?

Comment (diml) [reply test/blackbox-tests/test-cases/odoc/warn-error/run.t]: Well, by testing with an error in the source file, we are really testing the user-facing aspect of this PR, which is the most important.

If it's difficult to do that, we can keep the current way, but the output should be cleaned up more. Basically, a new reader should be able to look at the test and quickly understand what it is doing and whether it is passing or not. Otherwise, if some day this test breaks we might just accept the new output without realising the test is no longer passing.

Comment (Julow) [reply test/blackbox-tests/test-cases/odoc/warn-error/run.t]: I changed the tests to show the actual output. I agree it looks a bit better.

**Comment by diml** [comment]
To fix the CI failure, look for `odoc "odoc-simple"` in `test/blackbox-tests/gen_tests.ml` and add a line `odoc "odoc/warn-error"`.

I also prepared a PR to stop dune from printing command lines in tests (#3042). Once it's merged, you can rebase this PR and remove the `| sed ...` from the tests.

**Comment by diml** [comment]
> To fix the CI failure, look for odoc "odoc-simple" in test/blackbox-tests/gen_tests.ml and add a line odoc "odoc/warn-error".

And then run `make test` again. Basically this will setup things so that the new test is considered as a test that has external dependencies (odoc in this case).

**Comment by diml** [comment]
Alright, #3042 has just been merged

**Comment by Julow** [comment]
Just rebased and fixed the tests !

**Review by ghost — APPROVED** [review]
Looks good!

**Comment by diml** [comment]
Is the version of odoc that supports `--warn-error` released yet?

**Comment by Julow** [comment]
It's not yet

**Comment by diml** [comment]
Ok. Let's way for odoc to be released then, and then we can update the odoc version constraint in Dune and merge this PR.

File: src/dune/env_node.ml [hunk 4]
```diff
@@ -148,3 +154,24 @@ let rec foreign_flags t ~profile ~expander ~default_context_flags =
...
     t.foreign_flags <- Some flags;
     flags
+
+let rec odoc t ~profile =
+  match t.odoc with
+  | Some x -> x
+  | None ->
+    let open Odoc in
+    let default =
+      (* DUNE3: Enable for dev profile in the future *)
```
Comment (rgrinberg) [inline src/dune/env_node.ml]: This comment should apply only to the dev profile perhaps

**Comment by rgrinberg** [comment]
It's a bit inconsistent that we have semantic fields for odoc but general `<tool>_flags` fields for menhir  for example. Do you think it makes sense to add `odoc_flags` instead/in addition to this?

**Comment by rgrinberg** [comment]
This also needs an appropriate change list entry and a mention in the documentation.

**Comment by Julow** [comment]
I added a bit of doc.

Adding `odoc_flags` is not easy because Dune calls several Odoc sub commands (`html`, `compile` and `support-files`). These commands have a few options in common but would all interfere with the options provided by Dune (eg. `-I`, `-o`) or need to be passed only on a specific unit (`--hidden`).
Also, I think other options may be useful to configure later but will require logic in Dune (eg. `--syntax` could be passed automatically, `--theme-uri` should add a dependency on a file).

**Comment by rgrinberg** [comment]
I see. Thanks for the explanation.

File: doc/documentation.rst [hunk 5]
```diff
@@ -79,4 +79,21 @@ directory to the inferred package:
...
+.. code-block:: lisp
+
+    (env
+     (<profile>
+      (odoc <optional-fields>)))
+
+See :ref:`dune-env` for more details on the ``(env ...)`` stanza.
+``<optional-fields>`` are:
+
+- ``(warn-error <bool>)`` turns warnings into errors, if set to ``true``.
```
Comment (emillon) [inline doc/documentation.rst]: To clarify, can you document that the default is `false`?

File: doc/documentation.rst [hunk 6]
```diff
@@ -79,4 +79,21 @@ directory to the inferred package:
...
+  This field is available since Dune 2.2 and requires Odoc 1.4.3.
```
Comment (emillon) [inline doc/documentation.rst]: Also, I'm not too sure about the `warn-error` key. It mimics the configuration of the ocaml compilers, but that part is confusing.

What do you think about `(warnings fatal)` and `(warnings nonfatal)` (or maybe `(warnings display)`), which can be extended more easily to things like `(warnings silenced)` if we decide to extend the feature later?

File: src/dune/super_context.mli [hunk 7]
```diff
@@ -79,6 +79,9 @@ val menhir_flags :
     associated directory is [Path.relative dir ".bin"] *)
 val local_binaries : t -> dir:Path.Build.t -> File_binding.Expanded.t list
 
+(** Odoc's config the [env] stanza. *)
```
Comment (emillon) [inline src/dune/super_context.mli]:
```suggestion
(** odoc config in the corresponding [(env)] stanza. *)
```

File: src/dune_cache/dune_cache.ml [hunk 8]
```diff
@@ -282,12 +282,13 @@ module Cache = struct
     Io.write_file metadata_tmp_path metadata;
     let () =
       match Io.read_file metadata_path with
-        | contents -> if contents <> metadata then
```
Comment (emillon) [inline src/dune_cache/dune_cache.ml]: This hunk doesn't seem necessary

File: src/stdune/list.ml [hunk 9]
```diff
@@ -15,11 +15,10 @@ let is_non_empty = function
 let filter_map l ~f =
   let rec loop acc = function
     | [] -> rev acc
-    | x :: xs ->
-      begin match f x with
+    | x :: xs -> (
```
Comment (emillon) [inline src/stdune/list.ml]: same here, these hunks are not necessary

**Review by emillon — COMMENTED** [review]
I mostly had a look at the doc, thanks!

File: doc/documentation.rst (hunk 5 above)
Comment (Julow) [reply doc/documentation.rst]: Done

File: src/dune_cache/dune_cache.ml (hunk 8 above)
Comment (Julow) [reply src/dune_cache/dune_cache.ml]: Removed. It was due to formatting during rebase.

File: doc/documentation.rst (hunk 6 above)
Comment (Julow) [reply doc/documentation.rst]: Just did that. I agree it looks better.

**Comment by Julow** [comment]
Sorry for the delay, Odoc 1.5.0 has been release 12 days ago.
I made some changes after the last review.

**Comment by diml** [comment]
Could you update the odoc constraint in the dune-project file and .travis-ci.sh script, regenerate the opam file and rebase?

Then this should be good to merge.

**Comment by Julow** [comment]
I've update the PR, assuming the next release is `2.4.0`

**Comment by rgrinberg** [comment]
Could you rebase again? That should fix the CI failure.

File: src/stdune/list.ml [hunk 10]
```diff
@@ -15,10 +15,11 @@ let is_non_empty = function
 let filter_map l ~f =
   let rec loop acc = function
     | [] -> rev acc
-    | x :: xs -> (
-      match f x with
+    | x :: xs ->
+      begin match f x with
```
Comment (rgrinberg) [inline src/stdune/list.ml]: why was this changed?

**Comment by rgrinberg** [comment]
I've rebased the PR

File: src/stdune/list.ml (hunk 10 above)
Comment (Julow) [reply src/stdune/list.ml]: Sorry for the late reply. That file wasn't well formatted at the previous base of this branch (more than a month ago)

File: src/dune/odoc.ml [hunk 11]
```diff
@@ -252,8 +262,8 @@ let setup_html sctx (odoc_file : odoc) ~pkg ~requires =
           :: dune_keep ) )
 
 let setup_library_odoc_rules cctx (library : Library.t) ~dep_graphs =
+  let scope = Compilation_context.scope cctx in
   let lib =
-    let scope = Compilation_context.scope cctx in
```
Comment (rgrinberg) [inline src/dune/odoc.ml]: I've reverted this change
